Fix endblock panic for simulations where there are no bonded validators

adamewozniak opened the pull request:
Handles a case for simulations where there are no bonded validators

Closes #35

jkilpatr commented on module/x/gravity/abci.go:
```diff
@@ -35,7 +35,14 @@ func createValsets(ctx sdk.Context, k keeper.Keeper) {
 	significantPowerDiff := false
 	if latestValset != nil {
-		intCurrMembers, err := types.BridgeValidators(k.GetCurrentValset(ctx).Members).ToInternal()
+		vs, err := k.GetCurrentValset(ctx)
```
As discussed in my other comment, if we're making GetCurrentValset() return an error it should not panic and return two different types of errors, from there we can identify the empty valset one explicitly and log it here. I would also like to further guard for this by checking that we are in a simulation. Is there any way to do that?

adamewozniak replied:
I don't believe so, although I'll look into it

jkilpatr commented on module/x/gravity/keeper/keeper_valset.go:
```diff
@@ -251,8 +251,11 @@ func (k Keeper) IterateValsetBySlashedValsetNonce(ctx sdk.Context, lastSlashedVa
-func (k Keeper) GetCurrentValset(ctx sdk.Context) types.Valset {
+func (k Keeper) GetCurrentValset(ctx sdk.Context) (types.Valset, error) {
 	validators := k.StakingKeeper.GetBondedValidatorsByPower(ctx)
+	if len(validators) == 0 {
```
if this function is going to be made to return an error we should probably have it never panic and return specific errors for each case. Then we can identify that case and log instead of crashing in valset creation.

adamewozniak replied:
Wasn't sure if the other panics here were small enough issues to just log + error instead of panicking, but yeah makes sense 👍

ChristianBorst replied:
I think we do still want to panic in those other cases, but have the panic happen outside of this function by the caller.

jkilpatr replied:
Yup that sounds exactly aligned with what I'm thinking.

jkilpatr commented on module/x/gravity/keeper/querier.go:
```diff
@@ -31,7 +31,7 @@ func NewQuerier(keeper Keeper) sdk.Querier {
 
 func queryCurrentValset(ctx sdk.Context, keeper Keeper) ([]byte, error) {
-	valset := keeper.GetCurrentValset(ctx)
+	valset, _ := keeper.GetCurrentValset(ctx)
```
this should return an error if it receives one.

jkilpatr commented on module/x/gravity/keeper/grpc_query.go:
```diff
@@ -36,7 +36,8 @@ func (k Keeper) Params(c context.Context, req *types.QueryParamsRequest) (*types
 	c context.Context,
 	req *types.QueryCurrentValsetRequest) (*types.QueryCurrentValsetResponse, error) {
-	return &types.QueryCurrentValsetResponse{Valset: k.GetCurrentValset(sdk.UnwrapSDKContext(c))}, nil
+	vs, _ := k.GetCurrentValset(sdk.UnwrapSDKContext(c))
```
this should return the error if it receives one.

jkilpatr reviewed (changes requested):
This looks good. We still need a few more changes so that we can be 100% sure that we are not reducing the safety of regular operation.

ChristianBorst commented:
This looks good to me once @jkilpatr's suggestions are added.

adamewozniak commented:
> This looks good. We still need a few more changes so that we can be 100% sure that we are not reducing the safety of regular operation.

Pretty sure I addressed most of the concerns here, let me know if you have any more 👍

jkilpatr reviewed (approved):
This looks good to me, thanks for incorporating our suggestions!
 
Just waiting on tests to run before merging.

adamewozniak commented:
> This looks good to me, thanks for incorporating our suggestions!
> 
> Just waiting on tests to run before merging.

Looks good except for the relay-market test, which also failed for #39, but seems fine in `main` - let me know if you'd like me to look into it

jkilpatr commented:
Relay market is flaky don't think this is on your end. merging.
